### PR TITLE
[SPARK-39417][SQL] Handle Null partition values in PartitioningUtils

### PR DESCRIPTION
**Pull request by singhpk234**
### What changes were proposed in this pull request?

We should not try casting everything returned by `removeLeadingZerosFromNumberTypePartition` to string, as it returns null value for the cases when partition has null value and is already replaced by `DEFAULT_PARTITION_NAME`

### Why are the changes needed?

for null partitions where `removeLeadingZerosFromNumberTypePartition` is called it would throw a NPE and hence the query would fail. 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added a UT, which would fail with an NPE otherwise.

### CODE REVIEW

**Comment by singhpk234** [comment]
cc @srowen @HyukjinKwon

**Comment by huaxingao** [comment]
LGTM. Pending test results

**Review by HyukjinKwon — APPROVED** [review]

**Comment by huaxingao** [comment]
The Python doc generation failure seems to be irrelevant. All the other tests passed.

**Review by huaxingao — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
do we know which commit caused this issue? is it a 3.3 only bug?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 1]
```diff
@@ -359,7 +359,12 @@ object PartitioningUtils extends SQLConfHelper{
   def removeLeadingZerosFromNumberTypePartition(value: String, dataType: DataType): String =
     dataType match {
       case ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType =>
-        castPartValueToDesiredType(dataType, value, null).toString
+        val castedPartVal = castPartValueToDesiredType(dataType, value, null)
+        if (castedPartVal != null) {
+          castedPartVal.toString
+        } else {
+          null
+        }
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:367]: `Option(castPartValueToDesiredType(dataType, value, null)).map(_.toString).orNull`?

**Comment by singhpk234** [comment]
@cloud-fan, This seems to be introduced via [commit](https://github.com/apache/spark/commit/fc29c91f27d866502f5b6cc4261d4943b5cccc7e), [SPARK-35561](https://issues.apache.org/jira/browse/SPARK-35561) and seems to affect `3.3.0`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala [hunk 2]
```diff
@@ -1259,6 +1259,14 @@ class ParquetV2PartitionDiscoverySuite extends ParquetPartitionDiscoverySuite {
     assert("p_int=10/p_float=1.0" === path)
   }
 
+  test("SPARK-39417: Null partition value") {
+    // null partition value are replace by DEFAULT_PARTITION_NAME before hitting getPathFragment.
```
Comment (yaooqinn) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala:1263]: nit: `null partition value is replaced`

**Review by yaooqinn — APPROVED** [review]

**Review by wangyum — APPROVED** [review]

**Comment by huaxingao** [comment]
The python documentation failure is not related to this PR. All the other tests run successfully. I will merge this PR.

**Comment by huaxingao** [comment]
Merged to master/3.3. Thanks a lot @singhpk234 and et al.

**Comment by huaxingao** [comment]
@singhpk234 is this your first Spark contribution? I tried to assign the jira to you but somehow can't find you. Then I tried to add you to the Spark contributors list but found multiple users with the same name. Can you have your email?

**Pull request closed without merging**